File: java/src/edu/berkeley/cs/sort/Sort.java
package edu.berkeley.cs.sort;

public abstract class Sort<T extends Comparable> {
  private int swaps = 0;
  private int compares = 0;
  protected final int cutoff = 5;

  public abstract void sort(T[] input);

  public int swaps() {
    return swaps;
  }

  public int compares() {
    return compares;
  }

  protected boolean less(T v, T w) {
    compares++;
    return v.compareTo(w) < 0;
  }

  protected void swap(T[] input, int i, int j) {
    swaps++;
    T t = input[i];
    input[i] = input[j];
    input[j] = t;
  }
}

File: python/src/sort/sort.py
from abc import ABC, abstractmethod


class Sort(ABC):

    def __init__(self):
        self.swaps = 0
        self.compares = 0
        self.cutoff = 5

    @abstractmethod
    def sort(self, input):
        pass

    def less(self, v, w):
        self.compares += 1
        return v < w

    def swap(self, input, i, j):
        self.swaps += 1
        temp = input[i]
        input[i] = input[j]
        input[j] = temp
